### PR TITLE
top-level: Fix splicing, again

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

See line comments.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/splice.nix [hunk 1]
```diff
@@ -37,23 +37,21 @@ let
       inherit name;
       value = let
         defaultValue = mash.${name};
-        buildValue = buildPkgs.${name};
-        runValue = runPkgs.${name};
+        # `or {}` is for the non-derivation attsert splicing case, where `{}` is the identity.
```
Comment (Ericson2314) [inline pkgs/top-level/splice.nix]: So I was wrong, we did need some `or {}` for a different branch. `*Pkgs` should always evaluate to weak head normal form, so this is sufficient.

File: pkgs/top-level/splice.nix [hunk 2]
```diff
@@ -37,23 +37,21 @@ let
...
+        # evaluate we don't want to diverge the entire splice, so we fall back
+        # on {}
+        getOutputs = value0: let
+          inherit (builtins.tryEval value0) success value;
+        in lib.optionalAttrs success (lib.genAttrs
           (value.outputs or (lib.optional (value ? out) "out"))
-          (output: value.${output});
+          (output: value.${output}));
       in
-        # Certain *Cross derivations will fail assertions, but we need their
```
Comment (Ericson2314) [inline pkgs/top-level/splice.nix]: This was silly. If `defaultValue` diverges, `augmentedValue` will too.

File: pkgs/top-level/splice.nix [hunk 3]
```diff
@@ -37,23 +37,21 @@ let
...
-        # nativeDrv. We are assuming anything that fails to evaluate is an
-        # attrset (including derivation) and thus can be unioned.
-        if !(builtins.tryEval defaultValue).success then augmentedValue
         # The derivation along with its outputs, which we recur
         # on to splice them together.
         else if lib.isDerivation defaultValue then augmentedValue
```
Comment (orivej) [inline pkgs/top-level/splice.nix]: This `in else` won't evaluate!

Comment (Ericson2314) [reply pkgs/top-level/splice.nix]: Oops, thanks!

File: pkgs/top-level/splice.nix [hunk 4]
```diff
@@ -37,26 +37,25 @@ let
...
-        # Get the set of outputs of a derivation
+        # Get the set of outputs of a derivation. If one derivation fails to
+        # evaluate we don't want to diverge the entire splice, so we fall back
+        # on {}
         tryGetOutputs = value0: let
-          eval = builtins.tryEval value0;
-        in getOutputs (if eval.success then eval.value else {});
+          inherit (builtins.tryEval value0) success value;
+        in tryGetOutputs (lib.optionalAttrs success value)
         getOutputs = value: lib.genAttrs
```
Comment (orivej) [inline pkgs/top-level/splice.nix]: Missing semicolon before `getOutputs`.

File: pkgs/top-level/splice.nix [hunk 5]
```diff
@@ -37,26 +37,25 @@ let
...
           (value.outputs or (lib.optional (value ? out) "out"))
           (output: value.${output});
       in
-        # Certain *Cross derivations will fail assertions, but we need their
```
Comment (Ericson2314) [inline pkgs/top-level/splice.nix]: This old branch was silly. If `defaultValue` diverges, `augmentedValue` will too.

File: pkgs/top-level/splice.nix [hunk 6]
```diff
@@ -37,26 +37,25 @@ let
...
           // (lib.optionalAttrs (buildPkgs ? ${name}) { nativeDrv = buildValue; })
           // (lib.optionalAttrs (runPkgs ? ${name}) { crossDrv = runValue; });
-        # Get the set of outputs of a derivation
+        # Get the set of outputs of a derivation. If one derivation fails to
+        # evaluate we don't want to diverge the entire splice, so we fall back
+        # on {}
         tryGetOutputs = value0: let
-          eval = builtins.tryEval value0;
-        in getOutputs (if eval.success then eval.value else {});
+          inherit (builtins.tryEval value0) success value;
```
Comment (Ericson2314) [inline pkgs/top-level/splice.nix]: This is just a little code golf.

File: pkgs/top-level/splice.nix [hunk 7]
```diff
@@ -37,26 +37,25 @@ let
...
+        in tryGetOutputs (lib.optionalAttrs success value);
```
Comment (orivej) [inline pkgs/top-level/splice.nix]: Now `tryGetOutputs` is in infinite recursion.

**Pull request merged**